File: construct_postorder_from_in_and_pre.py
class TreeNode:
	def __init__(self, val):
		self.val = val
		self.left = None
		self.right = None


class Solution:
	def construct(self, preorder: str, inorder: str):
		# preorder empty -> return None
		if not preorder:
			return None

		root = TreeNode(preorder[0])
		inorder_idx = inorder.find(preorder[0])

		root.left = self.construct(preorder[1:inorder_idx + 1], inorder[:inorder_idx])
		root.right = self.construct(preorder[inorder_idx + 1:], inorder[inorder_idx + 1:])

		return root

	def postorder(self, node: TreeNode):
		if not node:
			return ""

		return self.postorder(node.left) + self.postorder(node.right) + node.val


def main():
	sol = Solution()
	preorder = "ABCDFEG"
	inorder = "BADFCEG"
	root = sol.construct(preorder, inorder)
	print(sol.postorder(root))

if __name__ == '__main__':
	main()

File: construct_postorder_from_in_and_pre.java
class Solution {
    private Map<Character, Integer> inorderMap;
    
    public TreeNode construct(String preorder, String inorder) {
        // 预处理：建立字符到索引的映射
        inorderMap = new HashMap<>();
        for (int i = 0; i < inorder.length(); i++) {
            inorderMap.put(inorder.charAt(i), i);
        }
        
        return buildTree(preorder, 0, preorder.length() - 1, 
                        inorder, 0, inorder.length() - 1);
    }
    
    private TreeNode buildTree(String preorder, int preStart, int preEnd,
                              String inorder, int inStart, int inEnd) {
        if (preStart > preEnd) return null;
        
        TreeNode root = new TreeNode(preorder.charAt(preStart));
        int inorderIdx = inorderMap.get(root.val); // O(1) 查找
        int leftSize = inorderIdx - inStart;
        
        root.left = buildTree(preorder, preStart + 1, preStart + leftSize,
                             inorder, inStart, inorderIdx - 1);
        root.right = buildTree(preorder, preStart + leftSize + 1, preEnd,
                              inorder, inorderIdx + 1, inEnd);
        
        return root;
    }
}